### ISSUE TITLE
Following .rustfmt.toml

### ISSUE THREAD

**Issue by AaronKutch** [andrewbaxter/genemichaels#45]
I'm liking the idea of `genemichaels` and especially how it automatically formats comments, but it doesn't look like it can be configured by settings like a `.rustfmt.toml` in a workspace. Something that isn't clear to me from the README, is a goal of `genemichaels` to have the same properties as `rustfmt` except it adds some hacks to handle some blocks more permissively?
Also, if you want a stress test my https://github.com/AaronKutch/awint library should help. It is failing on incorrectly thinking that complicated `@` bindings are tuples.

**Comment by andrewbaxter** [comment]
Ah, that's a good question. So it's kind of a grey answer, but while I had no specific goal to keep genemichaels and rustfmt formatting similar, when I had to make style decisions I typically deferred to rustfmt as a guide since I think its basic style is reasonable.  So the end result shouldn't be too jarringly different...

That said, I think the formatting algorithms are probably fundamentally different so any formatting options it has will probably be specific.  I can't comment much on rustfmt since I don't really know how it works, but genemichaels is a deterministic formatter that basically re-generates the source using only information from the AST, whereas I _think_ rustfmt tweaks the source using existing formatting as a guide.

I think the only real goal is to be 100% automatic, since I don't want to spend any time messing with formatting (anything, including comments :sweat_smile:) manually.

I'm happy to add more options if there's specific simple things people want to use and they don't require significant rewrites.

(And thanks for trying it out and the report! My number one priority is not crashing/breaking code, I'll take a look at that asap)

**Comment by andrewbaxter** [comment]
Okay, I fixed one issue with dropped comments in #46 .

I noticed there's some strange formatting with const impls:
```
impl < const BW: usize, const LEN: usize > const Deref for InlAwi < BW, LEN > {
    type Target = Bits;

    #[inline]
    fn deref(&self) -> &Self::Target {
        self.const_as_ref()
    }
}
```
I haven't dug in too deep, but the spaces around `<` suggest it's falling back to raw-token formatting which would mean that https://github.com/dtolnay/syn isn't parsing it. I don't see any indication that const impl isn't supported in that library yet, but it is an unstable feature I think?

I guess the @ bindings are in new work or a branch? I wasn't familiar with that syntax but it wouldn't surprise me if I missed a match variation, I'll write a few tests for it.

**Comment by andrewbaxter** [comment]
For the `@` bindings it seems to be nominally working when I test here: https://github.com/andrewbaxter/genemichaels/pull/49

Do you have an example of where it breaks?

**Comment by AaronKutch** [comment]
```
            match overall_alignment {
                FillerAlign::None => (),
                align @ (FillerAlign::Lsb | FillerAlign::Msb | FillerAlign::Mid) =>...
                ...
            }
```
It thinks the pattern after the `@` is a tuple and inserts a comma, resulting in broken compilation

**Comment by AaronKutch** [comment]
I think I would use `genemichaels` if there was a setting to only format comments. Here are some minor issues I have with it currently.
It formats
```
//! - `Bits` have a nonzero bit-width specified in a `NonZeroUsize`. Being
//!   nonzero, it eliminates several edge cases and ambiguities this crate would
//!   have to handle.
//! - `Bits` are stored in little endian order with the same requirements as
//!   `[usize]`. The number of `usize` digits is the minimum needed to store all
//!   bits. If bitwidth is not a multiple of `usize::BITS`, then there will be
//!   some unused bits in the last `usize` digit. For example, a bitwidth of of
//!   100 bits takes up 2 digits (if `usize::BITS == 64`): 64 bits in the first
//!   digit, 36 bits in the least significant bits of the second, and 28 unused
//!   bits in the remaining bits of the second.
//! - Unused bits are zeroed. Note that this is not a safety critical invariant.
//!   Setting unused bits via `Bits::as_mut_slice` or `Bits::last_mut` will not
//!   cause Rust U.B., but it may result in arithmetically incorrect results or
//!   panics from the functions on `Bits`. Arbitrary bits can in the last digit
//!   can be set temporarily, but [Bits::clear_unused_bits] should be run before
//!   reaching a function that expects all these invariants to hold.
```
into
```
//! * `Bits` have a nonzero bit-width specified in a `NonZeroUsize`. Being nonzero, it
//!    eliminates several edge cases and ambiguities this crate would have to handle.
//!
//! * `Bits` are stored in little endian order with the same requirements as`[usize]`. The
//!    number of `usize` digits is the minimum needed to store all bits. If bitwidth is not
//!    a multiple of `usize::BITS`, then there will be some unused bits in the last `usize`
//!    digit. For example, a bitwidth of of 100 bits takes up 2 digits (if `usize::BITS == 64`
//!    ): 64 bits in the first digit, 36 bits in the least significant bits of the second,
//!    and 28 unused bits in the remaining bits of the second.
//!
//! * Unused bits are zeroed. Note that this is not a safety critical invariant. Setting
//!    unused bits via `Bits::as_mut_slice` or `Bits::last_mut` will not cause Rust U.B.,
//!    but it may result in arithmetically incorrect results or panics from the functions on
//!    `Bits`. Arbitrary bits can in the last digit can be set temporarily, but
//!    [Bits::clear_unused_bits] should be run before reaching a function that expects all
//!    these invariants to hold.
```
There is a "as\n'[usize]'" squashed into "as'[usize]'" when it should be "as '[usize]'". Or in other words newlines get converted to spaces when wrapping changes. Stuff between two backticks should be treated as units that are "glued" together. Punctuation that immediately precedes or follows other punctuation or characters should be glued together (see how "(if 'usize::BITS == 64'):" gets broken with an awkward "):" on the next line, it could be broken up as "(if\n'usize::BITS == 64'):". This is more obvious with cases not shown here where lone periods get separated from the last word in their sentence.

I'm not sure if I should like or dislike my `-` blocks getting converted into `*` blocks. Maybe we should go with it since it is more visible in source code.

**Comment by andrewbaxter** [comment]
> ```
>             match overall_alignment {
>                 FillerAlign::None => (),
>                 align @ (FillerAlign::Lsb | FillerAlign::Msb | FillerAlign::Mid) =>...
>                 ...
>             }
> ```
> 
> It thinks the pattern after the `@` is a tuple and inserts a comma, resulting in broken compilation

Bleh, well `syn` is parsing it as a tuple:
```
/// A tuple pattern: `(a, b)`.
Tuple(PatTuple),
```
```
TUPLE (A :: B | A :: A)
```

I _think_ this isn't just incorrect documentation, since tuples can also appear in patterns (ex: `(el0, _)`) so it seems like a misparse.  I can probably just exclude the final comma here, but I'm not sure if that will break some unit...

Edit: yeah, just tested and
```
fn x((_z): (i32,)) {
    y(_z);
}
```
doesn't work.

**Comment by andrewbaxter** [comment]
https://github.com/dtolnay/syn/issues/1352

**Comment by andrewbaxter** [comment]
I think I can still work around this by detecting if there's an tuple with a single Pat::Or element and use custom logic, assuming this can't happen in other places.

**Comment by andrewbaxter** [comment]
Ah yeah, the inline-code followed by punctuation splitting is tricky.  I'll have to think about what I can do about it.  Something similar happened with text in links `[]` and I ended up hacking around it.  Probably a more general solution is needed.

**Comment by andrewbaxter** [comment]
I think I've fixed the first issue here: https://github.com/andrewbaxter/genemichaels/pull/63

**Comment by andrewbaxter** [comment]
Issue for comment wrapping: https://github.com/andrewbaxter/genemichaels/issues/48

**Comment by andrewbaxter** [comment]
Okay I _think_ I've fixed the wrapping near punctuation thing, although there may be some small issues left.  I made issues for having a config file and overriding markdown style, I might not get to those immediately though.

**Comment by AaronKutch** [comment]
Ok the comments are just about fixed besides the 87 length thing I mentioned in the other issue, and it would be nice if there were a setting to only format comments. Although I don't plan to use this part because I have very specific ways I want to format my macros, I should mention that the macros seem to have no limits:
```
#[macro_export]
macro_rules! unsafe_binop_for_each_mut{
    ($lhs: ident, $rhs: ident, $x: ident, $y: ident, $f: block, $clear_unused_bits: expr) => {
        if $lhs.bw() != $rhs.bw() {
            return None
        }
        else {
            unsafe {
                // Safety: This accesses all regular digits within their bounds. If the bitwidths are equal, then the
                // slice lengths are also equal.
                const_for!(i in {
                    0..$lhs.len()
                } {
                    let $x = $lhs.get_unchecked_mut(i);
                    let $y = $rhs.get_unchecked(i);
                    $f;
                });
            }
            if $clear_unused_bits {
                $lhs.clear_unused_bits()
            }
            Some(())
        }
    };
}
```
```
    test_nonequal_bw!(
        x0,
        x1,
        s0,
        s1
        // functions with signature `fn({ , &, &mut} self) -> ...`
        ;
        nzbw bw zero_ umax_ imax_ imin_ uone_ not_ is_zero is_umax is_imin is_uone lsb msb lz tz count_ones rev_ to_usize to_isize to_u8 to_i8 to_u16 to_i16 to_u32 to_i32 to_u64 to_i64 to_u128 to_i128 to_bool abs_;
        shl_ lshr_ ashr_ rotl_ rotr_
        // functions with signature `fn({ , &, &mut} self, rhs: { , &, &mut} Self) -> Option<...>`
        ;
        copy_ or_ and_ xor_ const_eq const_ne ult ule ugt uge ilt ile igt ige add_ sub_ rsb_
    );
```

**Comment by andrewbaxter** [comment]
I think comments are by default wrapped to 100 characters past the indentation, so the absolute length increases with indentation.  Otherwise you'd have comments getting unreadably short wrapping as you indented things.  I checked the comment above and it seems to be exactly 100 (possibly off by one, if my math is wrong).  You need to include the `// ` at the beginning though.

As for macros, I think basically there's no rule for wrapping undifferentiated space-separated sequences.  The break points are only at commas and semicolons.  I'm not sure making these a vertical list rather than a horizontal list would particularly improve things.

**Comment by andrewbaxter** [comment]
I made more specific issues for everything I think (and hopefully fixed the above comment length thing), and #71 for partial formatting in general.  I'm going to close this for now - if there's more things to discuss I think creating specific issues for those would be great.
